modules/phase_field/src/postprocessors/FeatureFloodCount.C

```diff
@@ -268,39 +268,139 @@ void FeatureFloodCount::communicateAndMerge()
+        mooseError("Error in _feature_sets sorting, map index: " << map_num);
+
+      feature_offset += _feature_counts_per_map[map_num];
+    }
+#endif
+  }
+}
+
+void
+FeatureFloodCount::buildLocalToGlobalIndices(std::vector<std::vector<unsigned int> > & local_to_global_all) const
+{
+  mooseAssert(processor_id() == 0, "This method must only be called on the root processor");
+
+  local_to_global_all.resize(_n_procs, std::vector<unsigned int>(_max_local_size));
+
+  for (decltype(_feature_sets.size()) i = 0, end_index = _feature_sets.size(); i < end_index; ++i)
+  {
+    // Get the local indices from the feature and build a map
+    for (const auto & local_index_pair : _feature_sets[i]._orig_ids)
+    {
+      mooseAssert(local_index_pair.first < _n_procs, local_index_pair.first << ", " << _n_procs);
+      mooseAssert(local_index_pair.second < _max_local_size, local_index_pair.second << ", " << _max_local_size);
+
+                              // rank                 // local index
+      local_to_global_all[local_index_pair.first][local_index_pair.second] = i;
+    }
+  }
+}
+
+void
+FeatureFloodCount::scatterIndices(std::vector<std::vector<unsigned int> > & local_to_global_all)
+{
+  // First send _feature_count and _max_local_size for sizing vectors
+
+  std::pair<unsigned int, unsigned int> sizes_pair(_feature_count, _max_local_size);
+  _communicator.broadcast(sizes_pair);
+
+  if (processor_id() == 0)
+  {
+    // TODO: Replace with MPI_SCATTER
+    mooseAssert(local_to_global_all.size() == _n_procs, "local_to_global not sized properly");
+
+    for (auto rank = decltype(_n_procs)(1); rank < _n_procs; ++rank)
+      _communicator.send(rank, local_to_global_all[rank]);
+  }
+  else // Ranks 1..n
+  {
+    // Set _feature_count and _max_local_size on remaining ranks
+    _feature_count = sizes_pair.first;
+    _max_local_size = sizes_pair.second;
+
+    // Now receive the local to global map from the root process
+    _local_to_global_feature_map.resize(_max_local_size);
+    _communicator.receive(0, _local_to_global_feature_map);
+  }
 }
 
 void
 FeatureFloodCount::finalize()
 {
   communicateAndMerge();
 
+  // local to global map (one per processor)
+  std::vector<std::vector<unsigned int> > local_to_global_all;
+  if (processor_id() == 0)
+    buildLocalToGlobalIndices(local_to_global_all);
+
+  scatterIndices(local_to_global_all);
+
+  if (processor_id() != 0)
+  {
+    /**
+     * The rest of the ranks just need to move their own data to the flat data structure.
+     * Note that sorting is unecessary. The original ID inside of the unmerged feature
+     * will contain the original position which can be used to look up the global id
+     * number later after we receive the map from the root process.
+     */
+    _feature_sets.reserve(_max_local_size);
+    for (auto & list_ref : _partial_feature_sets)
+      for (auto && feature : list_ref)
```

Comment:
I agree that it's not an rvalue reference, if it was then I wouldn't need the `std::move` below. However, I tried to compile this without the std::move and it didn't work. I guess it's a good thing that range based loops can't be rvalue references, since that doesn't really make sense. I can remove the double `&&`.